File: encode-decode/parity-play.py
def encode(message):
    binary = ''.join('{0:08b}'.format(ord(char), 'b') for char in message)
    return binary + '1' if binary.count('1') % 2 == 1 else binary + '0'
    

def decode(encoded_message):
    message = encoded_message[:-1]
    decoded_string = "".join(chr(int(binary, 2)) for binary in [message[i:i+8] for i in range(0, len(message), 8)])
    return decoded_string if message.count('1') % 2 == int(encoded_message[-1]) else "parity error"


def main():
    encoded = encode('hello')
    print(decode(encoded))

    encoded2 = encode("hello")
    encoded2 = encoded2[:-1] + '0'
    print(decode(encoded2))

main()

File: encode-decode/parity_play.cpp
#include <iostream>
#include <string>
#include <bitset>
#include <algorithm>
#include <sstream>
#include "parity_play.h"

using namespace std;

int main() {
    string str = "Hello World";
    string encoded = encode(str);
    cout << decode(encoded) << endl;
    
    encoded.replace(encoded.size() - 1, 1 , "T");     // change to incorrect base (SVR)
    cout << decode(encoded) << endl;
    return 0;
}

string encode(string str) {
    string binary = "";
    for (char& c : str) {
        binary += bitset<8>(c).to_string();
    }
    string parityByte = bitset<8>(count(binary.begin(), binary.end(), '1') % 255).to_string();
    binary += parityByte;
    return binaryToBase(binary);
}

string decode(string baseStr) {
    string str = "";
    string binary = baseToBinary(baseStr);
    for (int i = 0; i < binary.size() - 8; i+=8) {
        string byte = binary.substr(i, 8);
        str += char(bitset<8>(byte).to_ulong());
    }
    int parityCount = stoi(binary.substr(binary.size() - 8, 8), nullptr, 2);
    if (count(binary.begin(), binary.end() - 8, '1') == parityCount) {
        return str;
    } else {
        return "Parity error";
    }
}

string binaryToBase(string binary) {
    string baseStr = "";
    for (int i = 0; i < binary.size(); i+=2) {
        baseStr += binaryBase[binary.substr(i, 2)];
    }
    return baseStr;
}

string baseToBinary(string str) {
    string binary = "";
    for (int i = 0; i < str.size(); i++) {
        binary += baseBinary[str.substr(i, 1)];
    }
    return binary;
}


